File: CS/GrowthWare.Framework.BusinessData/BusinessLogicLayer/BGroups.cs
using GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces;
using GrowthWare.Framework.Common;
using GrowthWare.Framework.Model.Profiles;
using System;
using System.Collections;
using System.Data;
using System.Globalization;

namespace GrowthWare.Framework.BusinessData.BusinessLogicLayer
{
    /// <summary>
    /// Process business logic for accounts
    /// </summary>
    /// <remarks>
    /// <![CDATA[
    /// MSecurityEntityProfile can be found in the GrowthWare.Framework.ModelObjects namespace.  
    /// 
    /// The following properties are necessary for correct business logic operation.
    /// .ConnectionString
    /// .DALName
    /// .DALNameSpace
    /// ]]>
    /// </remarks>
    /// <example> This sample shows how to create an instance of the class.
    /// <code language="VB.NET">
    /// <![CDATA[
    /// Dim myBll as new BGroups(mySecurityEntityProfile)
    /// ]]>
    /// </code>
    /// </example>
    public class BGroups : BaseBusinessLogic
    {
        private IDGroups m_DGroups;

        /// <summary>
        /// Private BGroups() to ensure only new instances with passed parameters is used.
        /// </summary>
        /// <remarks></remarks>
        private BGroups()
        {
        }

        /// <summary>
        /// Parameters are need to pass along to the factory for correct connection to the desired datastore.
        /// </summary>
        /// <param name="securityEntityProfile">The Security Entity profile used to obtain the DAL name, DAL name space, and the Connection String</param>
        /// <param name="centralManagement">Boolean value indicating if the system is being used to manage multiple database instances.</param>
        /// <remarks></remarks>
        /// <example> This sample shows how to create an instance of the class.
        /// <code language="VB.NET">
        /// <![CDATA[
        /// MSecurityEntityProfile mSecurityEntityProfile = MSecurityEntityProfile = New MSecurityEntityProfile();
        /// mSecurityEntityProfile.Id = ConfigSettings.DefaultSecurityEntityID;
        /// mSecurityEntityProfile.DAL = ConfigSettings.DAL;
        /// mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL);
        /// mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL);
        /// mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString;
        /// 
        /// BGroupss mBGroups = BGroupss = New BGroupss(mSecurityEntityProfile, ConfigSettings.CentralManagement);
        /// ]]>
        /// </code>
        /// <code language="C#">
        /// <![CDATA[
        /// Dim mSecurityEntityProfile As MSecurityEntityProfile = New MSecurityEntityProfile()
        /// mSecurityEntityProfile.Id = ConfigSettings.DefaultSecurityEntityID
        /// mSecurityEntityProfile.DAL = ConfigSettings.DAL
        /// mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL)
        /// mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL)
        /// mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString
        /// 
        /// Dim mBGroups As BGroupss = New BGroupss(mSecurityEntityProfile, ConfigSettings.CentralManagement)
        /// ]]>
        /// </code>
        /// </example>
        public BGroups(MSecurityEntityProfile securityEntityProfile, bool centralManagement)
        {
            if (securityEntityProfile == null)
            {
                throw new ArgumentException("The securityEntityProfile and not be null!");
            }
            if (centralManagement)
            {
                if (m_DGroups == null)
                {
                    m_DGroups = (IDGroups)ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups");
                }
            }
            else
            {
                m_DGroups = (IDGroups)ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups");
            }

            m_DGroups.ConnectionString = securityEntityProfile.ConnectionString;
            m_DGroups.SecurityEntitySeqID = securityEntityProfile.Id;
        }

        /// <summary>
        /// Gets the groups by security entity.
        /// </summary>
        /// <param name="securityEntityId">The security entity ID.</param>
        /// <returns>DataTable.</returns>
        public DataTable GetGroupsBySecurityEntity(int securityEntityId)
        {
            MGroupProfile myProfile = new MGroupProfile();
            myProfile.SecurityEntityId = securityEntityId;
            m_DGroups.Profile = myProfile;
            return m_DGroups.GetGroupsBySecurityEntity();
        }

        /// <summary>
        /// Adds the group.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns>System.Int32.</returns>
        public void AddGroup(MGroupProfile profile)
        {
            m_DGroups.Profile = profile;
            m_DGroups.AddGroup();
        }

        /// <summary>
        /// Gets the profile.
        /// </summary>
        /// <param name="groupId">The group ID.</param>
        /// <returns>MGroupProfile.</returns>
        public MGroupProfile GetProfile(int groupId)
        {
            MGroupProfile retProfile = new MGroupProfile();
            retProfile.Id = groupId;
            m_DGroups.Profile = retProfile;
            retProfile = new MGroupProfile(m_DGroups.GetProfileData());
            return retProfile;
        }

        /// <summary>
        /// Deletes the group.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        public bool DeleteGroup(MGroupProfile profile)
        {
            bool retVal = false;
            m_DGroups.Profile = profile;
            retVal = m_DGroups.DeleteGroup();
            return retVal;
        }

        /// <summary>
        /// Updates the group.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        public void Save(MGroupProfile profile)
        {
            m_DGroups.Profile = profile;
            m_DGroups.Save();
        }

        /// <summary>
        /// Gets the selected roles.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns>System.String[][].</returns>
        public string[] GetSelectedRoles(MGroupRoles profile)
        {
            ArrayList ClientRoles = new ArrayList();
            m_DGroups.GroupRolesProfile = profile;
            DataTable myDataTable = m_DGroups.GetGroupRoles();
            DataRow myDR = null;
            foreach (DataRow myDR_loopVariable in myDataTable.Rows)
            {
                myDR = myDR_loopVariable;
                ClientRoles.Add(myDR["Role"].ToString());
            }
            return (string[])ClientRoles.ToArray(typeof(string));
        }

        /// <summary>
        /// Searches groups using the specified search criteria.
        /// </summary>
        /// <param name="searchCriteria">The search criteria.</param>
        /// <returns>DataTable.</returns>
        public DataTable Search( MSearchCriteria searchCriteria)
        {
            return m_DGroups.Search(searchCriteria);
        }

        /// <summary>
        /// Updates the group roles.
        /// </summary>
        /// <param name="profile">The profile.</param>
        /// <returns><c>true</c> if XXXX, <c>false</c> otherwise</returns>
        public bool UpdateGroupRoles(MGroupRoles profile)
        {
            m_DGroups.GroupRolesProfile = profile;
            return m_DGroups.UpdateGroupRoles();
        }
    }
}

File: VB/GrowthWare.Framework.BusinessData/BusinessLogicLayer/BGroups.vb
Imports GrowthWare.Framework.BusinessData.DataAccessLayer.Interfaces
Imports GrowthWare.Framework.Common
Imports GrowthWare.Framework.Model.Profiles

Namespace BusinessLogicLayer
    Public Class BGroups
        Private m_DGroups As IDGroups

        ''' <summary>
        ''' Private sub new() to ensure only new instances with passed parameters is used.
        ''' </summary>
        ''' <remarks></remarks>
        Private Sub New()

        End Sub

        Public Sub Save(ByVal profile As MGroupProfile)
            m_DGroups.Profile = profile
            m_DGroups.Save()
        End Sub

        ''' <summary>
        ''' Parameters are need to pass along to the factory for correct connection to the desired datastore.
        ''' </summary>
        ''' <param name="securityEntityProfile">The Security Entity profile used to obtain the DAL name, DAL name space, and the Connection String</param>
        ''' <param name="centralManagement">Boolean value indicating if the system is being used to manage multiple database instances.</param>
        ''' <remarks></remarks>
        ''' <example> This sample shows how to create an instance of the class.
        ''' <code language="VB.NET">
        ''' <![CDATA[
        ''' MSecurityEntityProfile mSecurityEntityProfile = MSecurityEntityProfile = New MSecurityEntityProfile();
        ''' mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID;
        ''' mSecurityEntityProfile.DAL = ConfigSettings.DAL;
        ''' mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL);
        ''' mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL);
        ''' mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString;
        ''' 
        ''' BGroups mBGroups = New BGroups(mSecurityEntityProfile, ConfigSettings.CentralManagement);
        ''' ]]>
        ''' </code>
        ''' <code language="C#">
        ''' <![CDATA[
        ''' Dim mSecurityEntityProfile As MSecurityEntityProfile = New MSecurityEntityProfile()
        ''' mSecurityEntityProfile.ID = ConfigSettings.DefaultSecurityEntityID
        ''' mSecurityEntityProfile.DAL = ConfigSettings.DAL
        ''' mSecurityEntityProfile.DAL_Namespace = ConfigSettings.DAL_NameSpace(mSecurityEntityProfile.DAL)
        ''' mSecurityEntityProfile.DAL_Name = ConfigSettings.DAL_AssemblyName(mSecurityEntityProfile.DAL)
        ''' mSecurityEntityProfile.ConnectionString = ConfigSettings.ConnectionString
        ''' 
        ''' Dim mBGroups As BGroups = New BGroups(mSecurityEntityProfile, ConfigSettings.CentralManagement)
        ''' ]]>
        ''' </code>
        ''' </example>
        Public Sub New(ByVal securityEntityProfile As MSecurityEntityProfile, ByVal centralManagement As Boolean)
            If securityEntityProfile Is Nothing Then Throw New ArgumentNullException("securityEntityProfile", "securityEntityProfile cannot be a null reference (Nothing in Visual Basic)!")
            If Not centralManagement Then
                If m_DGroups Is Nothing Then
                    m_DGroups = ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups")
                End If
            Else
                m_DGroups = ObjectFactory.Create(securityEntityProfile.DataAccessLayerAssemblyName, securityEntityProfile.DataAccessLayerNamespace, "DGroups")
            End If
            m_DGroups.ConnectionString = securityEntityProfile.ConnectionString
            m_DGroups.SecurityEntitySeqId = securityEntityProfile.Id
        End Sub

        Public Function GetGroupsBySecurityEntity(ByVal securityEntityId As Integer) As DataTable
            Dim myProfile As New MGroupProfile
            myProfile.SecurityEntityID = securityEntityId
            m_DGroups.Profile = myProfile
            Return m_DGroups.GroupsBySecurityEntity
        End Function

        Public Sub AddGroup(ByVal profile As MGroupProfile)
            m_DGroups.Profile = profile
            m_DGroups.Save()
        End Sub

        Public Function GetProfile(ByVal groupId As Integer) As MGroupProfile
            Dim retProfile As New MGroupProfile
            retProfile.ID = groupId
            m_DGroups.Profile = retProfile
            retProfile = New MGroupProfile(m_DGroups.ProfileData)
            Return retProfile
        End Function

        Public Sub DeleteGroup(ByVal profile As MGroupProfile)
            m_DGroups.Profile = profile
            m_DGroups.DeleteGroup()
        End Sub

        Public Sub UpdateGroup(ByVal profile As MGroupProfile)
            m_DGroups.Profile = profile
            m_DGroups.Save()
        End Sub

        ''' <summary>
        ''' Searches the specified search critera.
        ''' </summary>
        ''' <param name="searchCriteria">The search critera.</param>
        ''' <returns>DataTable.</returns>
        Public Function Search(ByVal searchCriteria As MSearchCriteria) As DataTable
            Return m_DGroups.Search(searchCriteria)
        End Function

        ''' <summary>
        ''' Gets the selected roles.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        ''' <returns>System.String[][].</returns>
        Public Function GetSelectedRoles(ByVal profile As MGroupRoles) As String()
            Dim ClientRoles As New ArrayList
            m_DGroups.GroupRolesProfile = profile
            Dim myDataTable As DataTable = m_DGroups.GroupRoles
            Dim myDR As DataRow
            For Each myDR In myDataTable.Rows
                ClientRoles.Add(myDR("Role").ToString)
            Next
            Return CType(ClientRoles.ToArray(GetType(String)), String())
        End Function

        ''' <summary>
        ''' Updates the group roles.
        ''' </summary>
        ''' <param name="profile">The profile.</param>
        Public Sub UpdateGroupRoles(ByVal profile As MGroupRoles)
            m_DGroups.GroupRolesProfile = profile
            m_DGroups.UpdateGroupRoles()
        End Sub
    End Class
End Namespace

